File: LeetCode/Python/leetcode_898_Bitwise_ORs_of_Subarrays.py
# LeetCode #898 - Bitwise ORs of Subarrays
# Date: 2025-07-31

# ************************ Brute Force ************************
class BruteSolution:
    def subarrayBitwiseORs(self, arr: List[int]) -> int:
        n = len(arr)
        s = set()
        for i in range(n):
            for j in range(i, n):
                curr = arr[i]
                for k in range(i, j + 1):
                    curr |= arr[k]
                s.add(curr)
        return len(s)

# ************************ Better ************************
class BetterSolution:
    def subarrayBitwiseORs(self, arr: List[int]) -> int:
        n = len(arr)
        s = set()
        for i in range(n):
            curr = 0
            for j in range(i, n):
                curr |= arr[j]
                s.add(curr)
        return len(s)

# ************************ Optimal ************************
class OptimalSolution:
    def subarrayBitwiseORs(self, arr: List[int]) -> int:
        n = len(arr)
        prev = set()
        curr = set()
        res = set()

        for i in range(n):
            for x in prev:
                curr.add(arr[i]|x)
                res.add(arr[i]|x)
            curr.add(arr[i])
            res.add(arr[i])
            prev = curr
            curr = set()
                
        return len(res)

File: LeetCode/C++/leetcode_898_Bitwise_ORs_of_Subarrays.cpp
// LeetCode #898 - Bitwise ORs of Subarrays
// Date: 2025-07-31

#include <bits/stdc++.h>
using namespace std;

// ************************ Brute Force ************************
class BruteSolution {
public:
    int subarrayBitwiseORs(vector<int>& arr) {
        int n = arr.size();
        set<int> s;
        for (int i = 0; i < n; i++) {
            for (int j = i; j < n; j++) {
                int curr = arr[i];
                for (int k = i; k <= j; k++) {
                    curr |= arr[k];
                }
                s.insert(curr);
            }
        }
        return s.size();
    }
};

// ************************ Better ************************
class BetterSolution {
public:
    int subarrayBitwiseORs(vector<int>& arr) {
        int n = arr.size();
        set<int> s;
        for (int i = 0; i < n; i++) {
            int curr = 0;
            for (int j = i; j < n; j++) {
                curr |= arr[j];
                s.insert(curr);
            }
        }
        return s.size();
    }
};

// ************************ Optimal ************************
class OptimalSolution {
public:
    int subarrayBitwiseORs(vector<int>& arr) {
        int n = arr.size();
        unordered_set<int> prev, curr, res;

        for (int i = 0; i < n; i++) {
            curr.clear();
            for (auto &x : prev) {
                curr.insert(arr[i] | x);
                res.insert(arr[i] | x);
            }
            curr.insert(arr[i]);
            res.insert(arr[i]);
            prev = curr;
            curr.clear();
        }

        return res.size();
    }
};
